### PR TITLE
Add yield signals to page processors

### PR DESCRIPTION
**Pull request by highker**
Wire yield signals into the iterator created by page processors. During
the computation of the next page, it may pause and return null pages.
The computation will continue when the iterator is called again. The
yield is at column processing granularity.

**Further splitting the PR into two (together with https://github.com/prestodb/presto/pull/8726) given that one would be rather complicated; let's move one step a time.**

### CODE REVIEW

File: presto-main/src/main/java/com/facebook/presto/operator/project/PageProcessor.java [hunk 1]
```diff
@@ -226,13 +248,20 @@ private void updateRetainedSize()
...
+                }
+
                 if (positionsBatch.size() > 1 && pageSize > MAX_PAGE_SIZE_IN_BYTES) {
                     return Optional.empty();
                 }
 
                 // if possible, use previouslyComputedResults produced in prior optimistic failure attempt
                 PageProjection projection = projections.get(i);
-                if (previouslyComputedResults[i] != null && previouslyComputedResults[i].getPositionCount() > batchSize) {
+                if (previouslyComputedResults[i] != null && previouslyComputedResults[i].getPositionCount() >= batchSize) {
```
Comment (findepi) [inline presto-main/src/main/java/com/facebook/presto/operator/project/PageProcessor.java]: Is the change `>` to `>=` related to yields?

File: presto-main/src/test/java/com/facebook/presto/operator/TestScanFilterAndProjectOperator.java [hunk 2]
```diff
@@ -181,6 +183,66 @@ public void testRecordCursorSource()
...
+                ImmutableList.of(),
+                ImmutableList.of(BIGINT));
+
+        SourceOperator operator = factory.createOperator(driverContext);
+        operator.addSplit(new Split(new ConnectorId("test"), TestingTransactionHandle.create(), TestingSplit.createLocalSplit()));
+        operator.noMoreSplits();
+
+        // we only check yield signal once a column has been completely processed; thus the first 19 outputs will get nothing
+        for (int i = 0; i < totalColumns - 1; i++) {
+            driverContext.getYieldSignal().setWithDelay(5 * SECONDS.toNanos(1), driverContext.getYieldExecutor());
```
Comment (findepi) [inline presto-main/src/test/java/com/facebook/presto/operator/TestScanFilterAndProjectOperator.java]: I saw this in `THJO.testYield` and found it misleading (and preventing any form of debugging). Wouldn't it be sufficient to let your `GenericLongFunction` force yield?

File: presto-main/src/main/java/com/facebook/presto/operator/project/PageProcessor.java (hunk 1 above)
Comment (highker) [inline presto-main/src/main/java/com/facebook/presto/operator/project/PageProcessor.java]: This is not; I should separate it into another commit

File: presto-main/src/test/java/com/facebook/presto/operator/TestScanFilterAndProjectOperator.java (hunk 2 above)
Comment (highker) [inline presto-main/src/test/java/com/facebook/presto/operator/TestScanFilterAndProjectOperator.java]: (1) `DriverYieldSignal` should be used with `setWithDelay` and `reset` alternating in order to yield, unyield, yield, unyield, ... There are several sanity checks to enforce that. That's why we need this `setWithDelay` here so we can `reset` later on. In unit tests, in order to precisely control the time to yield, we added the backdoor function: `forceYieldForTesting`. With the function, the `DriverYieldSignal` should be used with with `setWithDelay`, (`forceYieldForTesting`)+ and `reset` alternating. It is ok to put `forceYieldForTesting` inside `GenericLongFunction`; but we would love to have the unit test to fully control the logic.

Comment (findepi) [inline presto-main/src/test/java/com/facebook/presto/operator/TestScanFilterAndProjectOperator.java]: Yea, I know you can't just remove the line :) What about changing 5s to something closer to eternity?

File: presto-main/src/test/java/com/facebook/presto/operator/TestScanFilterAndProjectOperator.java [hunk 3]
```diff
@@ -181,6 +183,66 @@ public void testRecordCursorSource()
...
+            assertNull(operator.getOutput());
+            driverContext.getYieldSignal().reset();
+        }
+
+        // the last call will return the whole page
+        driverContext.getYieldSignal().setWithDelay(5 * SECONDS.toNanos(1), driverContext.getYieldExecutor());
```
Comment (dain) [inline presto-main/src/test/java/com/facebook/presto/operator/TestScanFilterAndProjectOperator.java]: just use `SECONDS.toNanos(5)`

Comment (dain) [inline presto-main/src/test/java/com/facebook/presto/operator/TestScanFilterAndProjectOperator.java]: I don't think this is needed
